Feature/sparkline chart

ChengYanJin opened the pull request:
**Component**: sparkline, linechart, vegachart

**Description**:
`sparkline`
- support light/dark mode
- add the max/min label 
- add the horizontal max/min line 

`vegachart && linechart`
- fix the color issue of the vertical line (in linechart if it's dark mode the color of vertical line is light)
- fix the tooltip theme 

**Design**:
light mode:
![image](https://user-images.githubusercontent.com/18453133/70228774-407fd580-1755-11ea-86ec-d4d24d34af6c.png)
linechart tooltip:
![image](https://user-images.githubusercontent.com/18453133/70245178-f2c79500-1775-11ea-8813-baa6e2eb8125.png)


dark mode:
![image](https://user-images.githubusercontent.com/18453133/70228788-4b3a6a80-1755-11ea-86cb-220f6fb22854.png)
linechart tooltip
![image](https://user-images.githubusercontent.com/18453133/70245220-0541ce80-1776-11ea-8ae7-13b1bed18b43.png)

Closes: #108 
Closes: #131

ChengYanJin committed:
sparkline: add sparkline based on vegachart

Refs: #108

ChengYanJin committed:
sparkline: add the test of sparkline

Refs: #108

ChengYanJin committed:
sparkline: add the title in vegachart in orger to change the theme

Refs: #108

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/core-ui/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




There is a conflict between your branch `feature/sparkline-chart` and the
destination branch `development/1.0`.

Please resolve the conflict on **the feature branch** (`feature/sparkline-chart`).

```bash
 $ git fetch
 $ git checkout origin/feature/sparkline-chart
 $ git merge origin/development/1.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:feature/sparkline-chart
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @carlito-scality

* @nicolas2bert

* @rachedbenmustapha

* @chengyanjin

* @monpote

ChengYanJin committed:
VegaChart: set the tooltip theme according to the current background color

Refs: #108

carlito-scality commented on src/lib/components/sparkline/SparkLine.component.js:
```diff
@@ -0,0 +1,129 @@
+  row,
+  color,
+  maxRuleColor = "#F1B434",
+  minRuleColor = "#006F62"
```
Use predefined colors from theme.js

carlito-scality commented on src/lib/components/sparkline/SparkLine.component.js:
```diff
@@ -0,0 +1,129 @@
+  minRuleColor = "#006F62"
+}: Props) {
+  const maxRuleColorLight = lighten(0.5, maxRuleColor).toString();
+  const minRuleColorLight = lighten(0.5, minRuleColor).toString();
```
we should not use `lighten` since  if `maxRuleColor` is already `light`, maxRuleColorLight will be white

ChengYanJin replied:
ok. So I will remove `maxRuleColorLight` and `minRuleColorLight` just use `opacity` which is prop in vega

carlito-scality commented on src/lib/components/sparkline/SparkLine.component.js:
```diff
@@ -0,0 +1,129 @@
+  row: Object,
+  color?: string,
+  maxRuleColor?: string,
+  minRuleColor?: string
```
you need add docs or comment to explain what the props are

ChengYanJin replied:
Rename with understandable name ;)

carlito-scality commented on src/lib/components/vegachart/VegaChart.component.js:
```diff
@@ -26,6 +26,13 @@ function VegaChart({ id, spec }: Props) {
+  } else {
+    tooltipTheme = "dark";
+  }
+
```
It is not a robust way. you may need to add `tooltipTheme ` to the props. There is another way to customize the tooltip instead of using vega default theme ?

ChengYanJin replied:
I added a new prop in `vegachart` component.

Actually, we can customize the theme for tooltip by override the CSS.  The only way that works for me is create a CSS file and import it in vegachart. 
I am wondering do you have better ideas?

Please find the following link that explains how to customize the tooltip by using `vega-tooltip`.
[doc - customize tooltip](https://github.com/vega/vega-tooltip/blob/master/docs/customizing_your_tooltip.md)

carlito-scality replied:
ok! according to the doc, `theme` could be anything (light, dark, custom...) So it is up to the user to come with their own CSS and configure `theme` to override the theme.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @carlito-scality

* @nicolas2bert

* @rachedbenmustapha

* @chengyanjin

* @monpote





The following reviewers are expecting changes from the author, or must review again:

* @carlito-scality

ChengYanJin committed:
sparkline: change the props name and add theme prop in vegachart

- Add a prop 'theme' in vegachart component which has two predefined option
in vega-tooltip: 'dark' and 'light'
- Rename the prop in sparkline to be more understandable

carlito-scality commented on src/lib/components/vegachart/VegaChart.component.js:
```diff
@@ -5,37 +5,62 @@ import vegaEmbed from "vega-embed";
   id: string,
-  spec: Object
+  spec: Object,
+  theme?: string // predefine theme light or dark
```
`theme` could be light, dark (vega predefine theme) and anything else (custom...)
Please put the link 'How to theme tooltip here'

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @carlito-scality

* @nicolas2bert

* @rachedbenmustapha

* @chengyanjin

* @monpote

ChengYanJin commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.0`




There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.0`




Please check the status of the associated issue None.

Goodbye chengyanjin.
